### PR TITLE
Allow lessees to create offers and leases

### PR DESCRIPTION
**Pull request by tzumainn**
This change allows lessees to create offers and leases within the duration
of their lease. If the parent lease is cancelled, the children offers and
leases will be cancelled as well.

### CODE REVIEW

File: esi_leap/db/sqlalchemy/api.py [hunk 1]
```diff
@@ -350,6 +354,56 @@ def lease_destroy(lease_uuid):
...
+               )
+    leases = add_lease_conflict_filter(leases, start, end)
+    lease_conflict = leases.first()
+
+    if lease_conflict:
+        raise exception.LeaseNoTimeAvailabilities(lease_uuid=lease_ref.uuid,
+                                                  start_time=start,
+                                                  end_time=end)
+
+    # check offer conflicts
```
[the remark this answers is not in the document]
Comment (tzumainn) [reply esi_leap/db/sqlalchemy/api.py:379]: Leases can be created outside of offers now, so both just represent segments of time when the resource is being used.

File: esi_leap/objects/lease.py [hunk 2]
```diff
@@ -138,16 +167,41 @@ def _fulfill_lease():
...
+            {'parent_lease_uuid': self.uuid,
+             'status': statuses.AVAILABLE},
+            None)
+        for offer in offers:
+            offer.expire(context)
+
+        @utils.synchronized(utils.get_resource_lock_name(self.resource_type,
+                                                         self.resource_uuid),
+                            external=True)
+        def _expire_lease():
```
Comment (DanNiESh) [inline esi_leap/objects/lease.py:186]: _expire_lease() and _cancel_lease() look similar. Can we take them out and make them one function?

Comment (tzumainn) [reply esi_leap/objects/lease.py:186]: Potentially! That would fall out of the scope of this PR however.

File: esi_leap/db/sqlalchemy/api.py [hunk 3]
```diff
@@ -156,7 +156,7 @@ def offer_get_conflict_times(offer_ref):
 
     return l_query.with_entities(
         models.Lease.start_time, models.Lease.end_time).\
-        join(models.Offer).\
+        join(models.Offer, models.Offer.uuid == models.Lease.offer_uuid).\
```
[the remark this answers is not in the document]
Comment (tzumainn) [reply esi_leap/db/sqlalchemy/api.py:159]: Ah, adding the new columns makes the join between Leases and Offers non-obvious to sqlalchemy, so it requires us to add this clause specifying how the join will take place.

**Review by DanNiESh — APPROVED** [review]
Looks good to me.

**Pull request merged**